Oidc additional client auth types

jkakavas opened the pull request:
The OpenID Connect specification defines a number of ways for a
client (RP) to authenticate itself to the OP when accessing the
Token Endpoint. We currently only support `client_secret_basic`.

This change introduces support for 2 additional authentication
methods, namely `client_secret_post` (where the client credentials
are passed in the body of the POST request to the OP) and
`client_secret_jwt` where the client constructs a JWT and signs
it using the the client secret as a key.

Support for the above, and especially `client_secret_jwt` in our
integration tests meant that the OP we use ( Connect2id server )
should be able to validate the JWT that we send it from the RP
(which contains the Token Endpoint as the `aud` claim ).
Since we run the OP in docker and it listens on an ephemeral port
we would have no way of knowing the port so that we can configure
the ES running via the testcluster to know the "correct" Token
Endpoint, and even if we did, this would not be the Token Endpoint
URL that the OP would think it listens on. To alleviate this, we
run an ES single node cluster in docker, alongside the OP so that
we can configured it with the correct hostname and port within
the docker network.

jkakavas committed:
wip

jkakavas committed:
OIDC: Enable additional client auth types

The OpenID Connect specification defines a number of ways for a
client (RP) to authenticate itself to the OP when accessing the
Token Endpoint. We currently only support `client_secret_basic`.

This change introduces support for 2 additional authentication
methods, namely `client_secret_post` (where the client credentials
are passed in the body of the POST request to the OP) and
`client_secret_jwt` where the client constructs a JWT and signs
it using the the client secret as a key.

Support for the above, and especially `client_secret_jwt` in our
integration tests meant that the OP we use ( Connect2id server )
should be able to validate the JWT that we send it from the RP.
Since we run the OP in docker and it listens on an ephemeral port
we would have no way of knowing the port so that we can configure
the ES running via the testcluster to know the "correct" Token
Endpoint, and even if we did, this would not be the Token Endpoint
URL that the OP would think it listens on. To alleviate this, we
run an ES single node cluster in docker, alongside the OP so that
we can configured it with the correct hostname and port within
the docker network.

jkakavas committed:
Merge remote-tracking branch 'origin/master' into oidc-additional-client-auth-types

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

jkakavas commented:
@mark-vieira added you for a +1 for the docker-compose changes since we chatted about this

jkakavas committed:
remove random characters from config

ywangd commented on x-pack/qa/oidc-op-tests/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthIT.java:
```diff
@@ -65,16 +71,31 @@
+        + "/c2id/clients";
+    private static final String LOGIN_API = "http://127.0.0.1:" + getEphemeralTcpPortFromProperty("oidc-provider", "8080")
+        + "/c2id-login/api/";
+    private static final String ES_PORT = getEphemeralTcpPortFromProperty("elasticsearch-node", "9200");
```
I don't see `test.fixtures.elasticsearch-node.9200` get set anywhere. How does this work?

jkakavas replied:
`elasticsearch-node` is the name of the elasticsearch service in https://github.com/elastic/elasticsearch/pull/58708/files#diff-2be63811028f90650408e594c29099b9R3

ywangd replied:
Ah awesome! buildSrc magic. TIL

ywangd reviewed (commented):
I still need go through the tests more carefully (they are new to me). Here are a few comments/questions to start. Thanks!

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java:
```diff
@@ -463,19 +466,36 @@ private void exchangeCodeForToken(AuthorizationCode code, ActionListener<Tuple<A
+            } else if (rpConfig.getClientAuthenticationMethod().equals(ClientAuthenticationMethod.CLIENT_SECRET_JWT)) {
+                ClientSecretJWT clientSecretJWT = new ClientSecretJWT(rpConfig.getClientId(), opConfig.getTokenEndpoint(),
+                    rpConfig.getClientAuthenticationJwtAlgorithm(), new Secret(rpConfig.getClientSecret().toString()));
+                for (Map.Entry<String, List<String>> entry : clientSecretJWT.toParameters().entrySet()) {
+                    // Both client_assertion and client_assertion_type are singleton lists
+                    params.add(new BasicNameValuePair(entry.getKey(), entry.getValue().get(0)));
+                }
```
> But client_id is not added as one of the request parameters?

It's not, it doesn't need to be. It is added as the `iss` claim of the JWT and validated from there. 

>  only contains client_assertion and client_assertion_type

yes, these are the two parameters needed, see https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

ywangd replied:
You are right. For a moment, I forgot how JWT is organized. It's too late for me ...

jkakavas replied:
I know that feeling !!! :) No need to go through this today, it can definitely wait. I raised the PR because I had it ready since last week

ywangd commented:
I assume this PR will close #54582 ? Or do you plan to add more auth methods after this one? Anyway, I am tagging it here for cross-reference.

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectRealmSettingsTests.java:
```diff
@@ -327,6 +362,45 @@ public void testInvalidProxyHostThrowsError() {
+            .put(getFullSettingKey(REALM_NAME, OpenIdConnectRealmSettings.PRINCIPAL_CLAIM.getClaim()), "sub")
+            .put(getFullSettingKey(REALM_NAME, OpenIdConnectRealmSettings.RP_REDIRECT_URI), "https://rp.my.com")
+            .put(getFullSettingKey(REALM_NAME, OpenIdConnectRealmSettings.RP_CLIENT_ID), "rp-my")
+            .put(getFullSettingKey(REALM_NAME, OpenIdConnectRealmSettings.RP_CLIENT_AUTH_METHOD), "none")
```
This would be testing the OP though, not our implementation, as it is the OP that should not validate a client JWT with none algorithm. We do _not_ disallow `none` specifically, we use an allowlist that only allows `"HS256", "HS384", "HS512"` 

I like the suggestion though, we can use it in `OpenIdConnectAuthenticatorTests.testImplicitFlowFailsWithNoneAlgorithm` which tests our implementation

jkakavas commented:
> I still need go through the tests more carefully (they are new to me). Here are a few comments/questions to start. Thanks!

we can wait until I return, this is not urgent, no worries.

jkakavas committed:
address feedback

mark-vieira commented on x-pack/test/idp-fixture/build.gradle:
```diff
@@ -1,4 +1,36 @@
+import org.elasticsearch.gradle.VersionProperties
+import org.elasticsearch.gradle.Architecture
+
 apply plugin: 'elasticsearch.build'
```
I wonder if we can remove `elasticsearch.build` plugin. This project contains no code or tests, it's simply a test fixture.

mark-vieira reviewed (approved):
One comment, otherwise LGTM.

ywangd reviewed (approved):
LGTM. I left a few questions about the c2id-server setup for my learning purpose. Thanks!

jkakavas commented on x-pack/qa/oidc-op-tests/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthIT.java:
```diff
@@ -95,30 +117,71 @@ public static void registerClients() throws Exception {
+            httpPost3.setEntity(new StringEntity(postClient, ContentType.APPLICATION_JSON));
+            httpPost3.setHeader("Accept", "application/json");
+            httpPost3.setHeader("Content-type", "application/json");
+            httpPost3.setHeader("Authorization", "Bearer 811fa888f3e0fdc9e01d4201bfeee46a");
```
We define this in override.properties which we bind mount to the container, see the docker compose file and it is used as a static credential to talk to the OPs API , I'll add a comment here when I get to merge this

jkakavas commented on x-pack/test/idp-fixture/docker-compose.yml:
```diff
@@ -40,11 +153,13 @@ services:
 
   oidc-provider:
-    image: "c2id/c2id-server:7.8"
+    image: "c2id/c2id-server:9.5"
```
Nope, we use the official one that the folks from connect2id ( who implement the OP ) publish in Dockerhub. I dont think they publish the Dockerfile, some details are in https://connect2id.com/products/server/docs/guides/docker

ywangd replied:
Thanks! I checked the image on Dockerhub and it has zero documentation. I found it hard to believe it was the official image ... apparently I am wrong.

jkakavas commented on x-pack/test/idp-fixture/oidc/override.properties:
```diff
@@ -1,4 +1,4 @@
-op.issuer=http://localhost:8080
-op.authz.endpoint=http://localhost:8080/c2id-login/
+op.issuer=http://oidc-provider:8080/c2id
+op.authz.endpoint=http://oidc-provider:8080/c2id-login/
```
See https://connect2id.com/products/server/docs/guides/docker for some details. Issuer is the OPs issuer string that it will use to populate the iss claim in the ID Token. The OP also uses this to determine the token  endpoint and userinfo endpoint it expects to be listening on by adding the proper suffix. 

The authz.endpoint is an API endpoint that we use to programmatically authenticate users with the OP.

We use the docker compose service name `oidc-provider` so that es which also runs in docker can access it via it (docker networking magic :) )

ywangd replied:
Sorry for not being clear here. I wasn't asking about `oidc-provider`, which I understand it's docker's doing. Was more about property names, e.g. `op.authz.endpoint` and what do they mean to the OP. The guide you link should answer all my questions. Thanks!

jkakavas committed:
Merge remote-tracking branch 'origin/master' into oidc-additional-client-auth-types

jkakavas committed:
address feedback

jkakavas commented:
@elasticmachine run elasticsearch-ci/2

jkakavas commented:
`:x-pack:test:idp-fixture:composeUp` fails for some reason, @breskeby is kindly assisting me pinpointing the cause

breskeby commented:
Hey looking at the `composeUp` issue I noticed two things:

1. As far as I understand the `composeUp` task should depend on `preProcessFixture` but it seems this is not wired up correctly. 
2. triggering  `preProcessFixture` might not fix the problem as it seems the up-to-date checking of this and its depending tasks returns false positives.

I need some more time to narrow down the proper fix. 


@jkakavas As a workaround, for now running `./gradlew preProcessFixture --rerun-tasks` triggers the preProcessFixture task and its dependencies with ignoring the `up-to-date` check and should fix the problem for you locally.

jkakavas commented:
@breskeby any updates on this ? Did you find sometime to look into the composeUp task? Anything I could do to help with this? Thanks!

breskeby commented:
@jkakavas I pushed a fix on task dependencies setup for test clusters earlier today. Can you rebase against master and see if that fixes your issue?

jkakavas commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into oidc-additional-client-auth-types

jkakavas commented:
@elasticmachine run elasticsearch-ci/packaging-sample-windows

jkakavas commented:
Still: 
```
http-proxy uses an image, skipping
oidc-provider uses an image, skipping
Creating network "89c62680e863be49b17213e549106bdb_idp-fixture__default" with the default driver
Pulling elasticsearch-node (elasticsearch:test)...
manifest for elasticsearch:test not found
Removing network 89c62680e863be49b17213e549106bdb_idp-fixture__default
```

@breskeby

breskeby commented:
@mark-vieira I looked briefly into this but couldn't find a root cause for this. do you have an idea what's going wrong here?

breskeby commented:
I saw a similar issue here: https://gradle-enterprise.elastic.co/s/dle4hccywpdic seems there's a general problem with the docker compose plugin

breskeby commented:
@jkakavas I pushed a potential fix to https://github.com/breskeby/elasticsearch/commit/de079d6dde022c5122a0c7b41ae9248abed4ad58 though what's not clear to me yet completely is how this test fixtures plugin is supposed to work. as mentioned earlier the composeUp in your case should dependOn `preProcessFixture` and I wonder if this is how things should be wired in general your if that specific scenario is special.

jkakavas commented:
I can verify this works. From my perspective, this is just blocking the merging of this PR as it doesn't allow to run the necessary tests to get a green CI. As such I can wait until we have a generic resolution ( given it's before 7.10 FF :) ) . I'll keep an eye out

jkakavas commented:
@ywangd would you mind merging this in my absence when the infra changes are in and ci/2 passes? If we determine that there is something that is wrong with my PR and needs to be addressed, I can take it up again when I'm back from PTO

ywangd commented:
> @ywangd would you mind merging this in my absence when the infra changes are in and ci/2 passes? If we determine that there is something that is wrong with my PR and needs to be addressed, I can take it up again when I'm back from PTO

Happy to keep watch on this one.

jkakavas commented:
Removed the `Resolves: #54582` as there are still client authentication types that we don't support ( namely `private_key_jwt` ) and that we can tackle in followups

mapdegree commented:
Hi @jkakavas ,

Would be very much appreciated, if ElasticSearch 7.9.x could support `client_secret_jwt` instead of only plain `secret-id`.

We use security hardened configuration o the Ping Identity OIDC provider and cannot deviate using only secret-id unfortunately.

ywangd commented:
> Would be very much appreciated, if ElasticSearch 7.9.x could support `client_secret_jwt` instead of only plain `secret-id`.
> 
The PR targets `v7.10.0`. Also patch releases, e.g. `v7.9.2`, normally only get bug fixes. Since this PR is new features, it is unlikely to be part of `v7.9.2`. But I'll double confirm with the team and get back to you.

ywangd committed:
Merge remote-tracking branch 'origin/master' into oidc-additional-client-auth-types

ywangd committed:
Apply Rene's fix for test fixture

From
https://github.com/breskeby/elasticsearch/commit/de079d6dde022c5122a0c7b41ae9248abed4ad58

ywangd commented:
Backport #62289
The change should be part of the upcoming `v7.10.0` release.

tvernum commented:
@jkakavas @ywangd Can we check the labels here?
I think we need to add `v8.0.0` and remove `backport pending`

ywangd commented:
@tvernum You are right. They are corrected now.
